src/frontend/Ast_to_Mir.ml

```diff
@@ -97,60 +97,124 @@ let neg_inf =
+  let inclusive_bound tp (lb : Expr.Typed.t) =
     if UnsizedType.is_int_type tp then
-      Ast.
-        { emeta
-        ; expr= BinOp (lb, Operator.Minus, {emeta; expr= Ast.IntNumeral "1"}) }
+      Expr.Helpers.binop lb Minus Expr.Helpers.one
     else lb in
+  let size_adjust e =
+    if
+      (not (UnsizedType.is_container ast_obs.Ast.emeta.type_))
+      || List.exists
+           ~f:(fun (i : Ast.typed_expression) ->
+             UnsizedType.is_container i.emeta.type_ )
+           ast_args
+    then e
+    else
+      (* Container y but scalar args - need to multiply by size(y) *)
+      let trans_ast_obs = trans_expr ast_obs in
+      let type_ = {trans_ast_obs.meta with type_= UnsizedType.UReal} in
+      Expr.Helpers.binop e Times
+        (Expr.Helpers.internal_funapp FnLength [trans_ast_obs] type_) in
   match t with
   | Ast.NoTruncate -> []
   | TruncateUpFrom lb ->
       let fk, fn, tp = find_function_info ccdf_suffices in
-      [ trunc Less lb
-          (targetme lb.emeta.loc
-             (funapp lb.emeta fk fn (inclusive_bound tp lb :: ast_args)) ) ]
+      let lb = trans_expr lb in
+      [ trunc Less "min" lb
+          (targetme lb.meta.loc
+             (size_adjust
+                (funapp lb.meta fk fn
+                   (inclusive_bound tp lb :: trans_exprs ast_args) ) ) ) ]
   | TruncateDownFrom ub ->
       let fk, fn, _ = find_function_info cdf_suffices in
-      [ trunc Greater ub
-          (targetme ub.emeta.loc (funapp ub.emeta fk fn (ub :: ast_args))) ]
+      let ub = trans_expr ub in
+      [ trunc Greater "max" ub
+          (targetme ub.meta.loc
+             (size_adjust (funapp ub.meta fk fn (ub :: trans_exprs ast_args))) )
+      ]
   | TruncateBetween (lb, ub) ->
       let fk, fn, tp = find_function_info cdf_suffices in
-      [ trunc Less lb
-          (trunc Greater ub
-             (targetme ub.emeta.loc
-                (funapp ub.emeta (Ast.StanLib FnPlain) "log_diff_exp"
-                   [ funapp ub.emeta fk fn (ub :: ast_args)
-                   ; funapp ub.emeta fk fn (inclusive_bound tp lb :: ast_args)
-                   ] ) ) ) ]
+      let lb, ub = (trans_expr lb, trans_expr ub) in
+      let expr args =
+        funapp ub.meta (Ast.StanLib FnPlain) "log_diff_exp"
+          [ funapp ub.meta fk fn (ub :: args)
+          ; funapp ub.meta fk fn (inclusive_bound tp lb :: args) ] in
+      let statement =
+        match
+          List.find
+            ~f:(fun (i : Ast.typed_expression) ->
+              UnsizedType.is_container i.emeta.type_ )
+            ast_args
+        with
+        (* If any of the arguments (besides the data) are vectors, need to generate a loop
+           This can go away if https://github.com/stan-dev/stan/issues/1154 is implemented
+        *)
+        | Some e ->
+            let e = trans_expr e in
+            let bound =
+              Expr.Helpers.internal_funapp FnLength [e]
+                {e.meta with type_= UnsizedType.UInt} in
```

Comment:
You should use the temporary for the loop size. I'm seeing generated code like
```cpp
for (int sym3__ = 1;
     sym3__ <= stan::math::size(
                 stan::math::add(stan::math::multiply(a, x), b));
     ++sym3__) {
```

Reply:
Fixed